### ISSUE TITLE
Completion handler may be invoked with an int instead of size_t, causing narrowing conversion

### ISSUE THREAD

**Issue by fabiorossetto** [boostorg/beast#2766]
### Version of Beast
1.83.0

### Problem description
In `impl/basic_stream.hpp` we find the following code:

```c++
        this->set_allowed_cancellation(net::cancellation_type::all);
        if (buffer_bytes(b_) == 0 && state().pending)
        {
            this->complete(false, error_code(), 0);
        }
        else
        {
            pg_.assign(state().pending);
            (*this)({});
        }
```

Passing "0" as size will cause the completion handler to be invoked with an `int` for the size. This can lead to narrowing conversions.

Simplest example:
```c++
#include <boost/beast/core/tcp_stream.hpp>

using namespace boost::beast;

struct Outcome {
   boost::system::error_code code;
   size_t size;
};

int main() {
    boost::asio::io_context context;
    tcp_stream stream{context};
    std::vector<char> buffer;
    stream.async_read_some(boost::asio::buffer(buffer), []<typename... Args>(Args&&... args){
        Outcome{std::forward<Args>(args)...};
    });
}
```

With `-Wall -Werror`, the code does not compile.

**Comment by ashtum** [comment]
@fabiorossetto That's correct; the second argument should have an explicit type of `std::size_t`. I believe we would be able to fix it within the time window for bug fixes in Boost 1.84.

**Comment by fabiorossetto** [comment]
Wow that was fast! Thank you!

**Comment by ashtum** [comment]
The fix has been merged into the master branch and will be included in the Boost 1.84 release.
https://github.com/boostorg/beast/blob/d0dd9c50694bcbb8836775e780eb046f66b3e1d7/include/boost/beast/core/impl/basic_stream.hpp#L300
